# Google data analytics Capstone Project: How does a bike-share navigate speedy success

Author: **[Ladislau André](http://https://ladislau2020.github.io/2021/CV/index.html)**
dataset: [Cyclistic Dataset](http://)
Date: *04-06-2022*



## Introduction


### Loading the libraries
We will be importing the pandas library for the ETL process.

In [1]:
import pandas as pd 

### Loading the Dataset

As earlier said we are going to be using the cyclist dataset for the last 12 months for this project.

In [2]:

month1 = pd.read_csv("../input/cyclistic-bike-share/202004-divvy-tripdata.csv")
month2 = pd.read_csv("../input/cyclistic-bike-share/202005-divvy-tripdata.csv")
month3 = pd.read_csv("../input/cyclistic-bike-share/202006-divvy-tripdata.csv")
month4 = pd.read_csv("../input/cyclistic-bike-share/202007-divvy-tripdata.csv")
month5 = pd.read_csv("../input/cyclistic-bike-share/202008-divvy-tripdata.csv")
month6 = pd.read_csv("../input/cyclistic-bike-share/202009-divvy-tripdata.csv")
month7 = pd.read_csv("../input/cyclistic-bike-share/202010-divvy-tripdata.csv")
month8 = pd.read_csv("../input/cyclistic-bike-share/202011-divvy-tripdata.csv")
month9 = pd.read_csv("../input/cyclistic-bike-share/202012-divvy-tripdata.csv")
month10 = pd.read_csv("../input/cyclistic-bike-share/202101-divvy-tripdata.csv")
month11 = pd.read_csv("../input/cyclistic-bike-share/202102-divvy-tripdata.csv")
month12 = pd.read_csv("../input/cyclistic-bike-share/202103-divvy-tripdata.csv")

# after loading the dataset we are going to bind them using the concat function

tripdata = pd.concat([month1, month2, month3, month4, month5, month6, month7, month8, month9, month10, month11, month12])


### Inspecting my dataset and gain insights about it

I will be using functions allowing me to inspect the dataset and get to know more about it.

In [3]:
print(tripdata)

                 ride_id  rideable_type           started_at  \
0       A847FADBBC638E45    docked_bike  2020-04-26 17:45:14   
1       5405B80E996FF60D    docked_bike  2020-04-17 17:08:54   
2       5DD24A79A4E006F4    docked_bike  2020-04-01 17:54:13   
3       2A59BBDF5CDBA725    docked_bike  2020-04-07 12:50:19   
4       27AD306C119C6158    docked_bike  2020-04-18 10:22:59   
...                  ...            ...                  ...   
228491  9397BDD14798A1BA    docked_bike  2021-03-20 14:58:56   
228492  BBBEB8D51AAD40DA   classic_bike  2021-03-02 11:35:10   
228493  637FF754DA0BD9E1   classic_bike  2021-03-09 11:07:36   
228494  F8F43A0B978A7A35   classic_bike  2021-03-01 18:11:57   
228495  3AE64EA5BF43CF72  electric_bike  2021-03-26 17:58:14   

                   ended_at            start_station_name start_station_id  \
0       2020-04-26 18:12:03                  Eckhart Park               86   
1       2020-04-17 17:17:03     Drake Ave & Fullerton Ave              503 

We have merged the whole data and we could notice that it has [3489748 rows x 13 columns] 

In [4]:
#Watching the first five rows of the data for inspecting purposes

tripdata.head()

ride_id rideable_type           started_at             ended_at  \
0  A847FADBBC638E45   docked_bike  2020-04-26 17:45:14  2020-04-26 18:12:03   
1  5405B80E996FF60D   docked_bike  2020-04-17 17:08:54  2020-04-17 17:17:03   
2  5DD24A79A4E006F4   docked_bike  2020-04-01 17:54:13  2020-04-01 18:08:36   
3  2A59BBDF5CDBA725   docked_bike  2020-04-07 12:50:19  2020-04-07 13:02:31   
4  27AD306C119C6158   docked_bike  2020-04-18 10:22:59  2020-04-18 11:15:54   

             start_station_name start_station_id             end_station_name  \
0                  Eckhart Park               86  Lincoln Ave & Diversey Pkwy   
1     Drake Ave & Fullerton Ave              503              Kosciuszko Park   
2          McClurg Ct & Erie St              142   Indiana Ave & Roosevelt Rd   
3  California Ave & Division St              216       Wood St & Augusta Blvd   
4          Rush St & Hubbard St              125   Sheridan Rd & Lawrence Ave   

  end_station_id  start_lat  start_lng  end_lat  end_lng member_casual  
0          152.0    41.8964   -87.6610  41.9322 -87.6586        member  
1          499.0    41.9244   -87.7154  41.9306 -87.7238        member  
2          255.0    41.8945   -87.6179  41.8679 -87.6230        member  
3          657.0    41.9030   -87.6975  41.8992 -87.6722        member  
4          323.0    41.8902   -87.6262  41.9695 -87.6547        casual

In [5]:
#The main informations about my dataset as max, min and others

tripdata.describe()

start_lat     start_lng       end_lat       end_lng
count  3.489748e+06  3.489748e+06  3.485010e+06  3.485010e+06
mean   4.190417e+01 -8.764494e+01  4.190444e+01 -8.764522e+01
std    4.364222e-02  2.575969e-02  4.373705e-02  2.589123e-02
min    4.164000e+01 -8.787000e+01  4.154000e+01 -8.807000e+01
25%    4.188224e+01 -8.765888e+01  4.188266e+01 -8.765917e+01
50%    4.190000e+01 -8.764170e+01  4.190068e+01 -8.764275e+01
75%    4.193000e+01 -8.762773e+01  4.193120e+01 -8.762775e+01
max    4.208000e+01 -8.752000e+01  4.216000e+01 -8.744000e+01

In [6]:
# This function shows random rows informations or observations

tripdata.tail()

ride_id  rideable_type           started_at  \
228491  9397BDD14798A1BA    docked_bike  2021-03-20 14:58:56   
228492  BBBEB8D51AAD40DA   classic_bike  2021-03-02 11:35:10   
228493  637FF754DA0BD9E1   classic_bike  2021-03-09 11:07:36   
228494  F8F43A0B978A7A35   classic_bike  2021-03-01 18:11:57   
228495  3AE64EA5BF43CF72  electric_bike  2021-03-26 17:58:14   

                   ended_at        start_station_name start_station_id  \
228491  2021-03-20 17:22:47     Michigan Ave & Oak St            13042   
228492  2021-03-02 11:43:37  Kingsbury St & Kinzie St     KA1503000043   
228493  2021-03-09 11:49:11     Michigan Ave & Oak St            13042   
228494  2021-03-01 18:18:37  Kingsbury St & Kinzie St     KA1503000043   
228495  2021-03-26 18:06:43                       NaN              NaN   

             end_station_name end_station_id  start_lat  start_lng    end_lat  \
228491   New St & Illinois St   TA1306000013  41.900960 -87.623777  41.890847   
228492   New St & Illinois St   TA1306000013  41.889177 -87.638506  41.890847   
228493  Clark St & Berwyn Ave   KA1504000146  41.900960 -87.623777  41.977997   
228494   New St & Illinois St   TA1306000013  41.889177 -87.638506  41.890847   
228495   New St & Illinois St   TA1306000013  41.880000 -87.630000  41.890818   

          end_lng member_casual  
228491 -87.618617        casual  
228492 -87.618617        member  
228493 -87.668047        member  
228494 -87.618617        member  
228495 -87.619540        member

In [7]:
# More information about the data, this functions displays the type of the variables or columns
# it also shows the count of the data format

tripdata.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3489748 entries, 0 to 228495
Data columns (total 13 columns):
 #   Column              Dtype  
---  ------              -----  
 0   ride_id             object 
 1   rideable_type       object 
 2   started_at          object 
 3   ended_at            object 
 4   start_station_name  object 
 5   start_station_id    object 
 6   end_station_name    object 
 7   end_station_id      object 
 8   start_lat           float64
 9   start_lng           float64
 10  end_lat             float64
 11  end_lng             float64
 12  member_casual       object 
dtypes: float64(4), object(9)
memory usage: 372.7+ MB


**Now that we have a quick overview of what our dataframe looks like, we will clean the data and prepare for analysis**

# Data Cleaning Process
**Remove duplicated rows using the ride_id variable**



In [8]:
# Using the duplicated() function combined with the sum() to have the total number of duplicated values

tripdata["ride_id"].duplicated().sum()
 

209

In [9]:
tripdata["ride_id"].value_counts() > 1

# we could see that there are duplicated values

EAB41740602B3BA6     True
FF49C6C596507F0B     True
87D4C892CA85DB8F     True
86E0216311A75D74     True
50031BC4290040DF     True
                    ...  
2200D3DB88D6541F    False
0DBF9C9E5B2D15D7    False
939883F8F15CF27B    False
2B05690EF1C97E8B    False
3AE64EA5BF43CF72    False
Name: ride_id, Length: 3489539, dtype: bool

In [10]:
# assigning the filtered data to a new variable with the duplicates removed 

tripdataV1 = tripdata.drop_duplicates(subset=["ride_id"])

tripdataV1

# It´s good to know that i have made the difference between the tripdata and the tripdata_filtered to have sure we removed the exact number of duplicated rows and yeah it was 209.

ride_id  rideable_type           started_at  \
0       A847FADBBC638E45    docked_bike  2020-04-26 17:45:14   
1       5405B80E996FF60D    docked_bike  2020-04-17 17:08:54   
2       5DD24A79A4E006F4    docked_bike  2020-04-01 17:54:13   
3       2A59BBDF5CDBA725    docked_bike  2020-04-07 12:50:19   
4       27AD306C119C6158    docked_bike  2020-04-18 10:22:59   
...                  ...            ...                  ...   
228491  9397BDD14798A1BA    docked_bike  2021-03-20 14:58:56   
228492  BBBEB8D51AAD40DA   classic_bike  2021-03-02 11:35:10   
228493  637FF754DA0BD9E1   classic_bike  2021-03-09 11:07:36   
228494  F8F43A0B978A7A35   classic_bike  2021-03-01 18:11:57   
228495  3AE64EA5BF43CF72  electric_bike  2021-03-26 17:58:14   

                   ended_at            start_station_name start_station_id  \
0       2020-04-26 18:12:03                  Eckhart Park               86   
1       2020-04-17 17:17:03     Drake Ave & Fullerton Ave              503   
2       2020-04-01 18:08:36          McClurg Ct & Erie St              142   
3       2020-04-07 13:02:31  California Ave & Division St              216   
4       2020-04-18 11:15:54          Rush St & Hubbard St              125   
...                     ...                           ...              ...   
228491  2021-03-20 17:22:47         Michigan Ave & Oak St            13042   
228492  2021-03-02 11:43:37      Kingsbury St & Kinzie St     KA1503000043   
228493  2021-03-09 11:49:11         Michigan Ave & Oak St            13042   
228494  2021-03-01 18:18:37      Kingsbury St & Kinzie St     KA1503000043   
228495  2021-03-26 18:06:43                           NaN              NaN   

                   end_station_name end_station_id  start_lat  start_lng  \
0       Lincoln Ave & Diversey Pkwy          152.0  41.896400 -87.661000   
1                   Kosciuszko Park          499.0  41.924400 -87.715400   
2        Indiana Ave & Roosevelt Rd          255.0  41.894500 -87.617900   
3            Wood St & Augusta Blvd          657.0  41.903000 -87.697500   
4        Sheridan Rd & Lawrence Ave          323.0  41.890200 -87.626200   
...                             ...            ...        ...        ...   
228491         New St & Illinois St   TA1306000013  41.900960 -87.623777   
228492         New St & Illinois St   TA1306000013  41.889177 -87.638506   
228493        Clark St & Berwyn Ave   KA1504000146  41.900960 -87.623777   
228494         New St & Illinois St   TA1306000013  41.889177 -87.638506   
228495         New St & Illinois St   TA1306000013  41.880000 -87.630000   

          end_lat    end_lng member_casual  
0       41.932200 -87.658600        member  
1       41.930600 -87.723800        member  
2       41.867900 -87.623000        member  
3       41.899200 -87.672200        member  
4       41.969500 -87.654700        casual  
...           ...        ...           ...  
228491  41.890847 -87.618617        casual  
228492  41.890847 -87.618617        member  
228493  41.977997 -87.668047        member  
228494  41.890847 -87.618617        member  
228495  41.890818 -87.619540        member  

[3489539 rows x 13 columns]

In [11]:

tripdataV1

        

ride_id  rideable_type           started_at  \
0       A847FADBBC638E45    docked_bike  2020-04-26 17:45:14   
1       5405B80E996FF60D    docked_bike  2020-04-17 17:08:54   
2       5DD24A79A4E006F4    docked_bike  2020-04-01 17:54:13   
3       2A59BBDF5CDBA725    docked_bike  2020-04-07 12:50:19   
4       27AD306C119C6158    docked_bike  2020-04-18 10:22:59   
...                  ...            ...                  ...   
228491  9397BDD14798A1BA    docked_bike  2021-03-20 14:58:56   
228492  BBBEB8D51AAD40DA   classic_bike  2021-03-02 11:35:10   
228493  637FF754DA0BD9E1   classic_bike  2021-03-09 11:07:36   
228494  F8F43A0B978A7A35   classic_bike  2021-03-01 18:11:57   
228495  3AE64EA5BF43CF72  electric_bike  2021-03-26 17:58:14   

                   ended_at            start_station_name start_station_id  \
0       2020-04-26 18:12:03                  Eckhart Park               86   
1       2020-04-17 17:17:03     Drake Ave & Fullerton Ave              503   
2       2020-04-01 18:08:36          McClurg Ct & Erie St              142   
3       2020-04-07 13:02:31  California Ave & Division St              216   
4       2020-04-18 11:15:54          Rush St & Hubbard St              125   
...                     ...                           ...              ...   
228491  2021-03-20 17:22:47         Michigan Ave & Oak St            13042   
228492  2021-03-02 11:43:37      Kingsbury St & Kinzie St     KA1503000043   
228493  2021-03-09 11:49:11         Michigan Ave & Oak St            13042   
228494  2021-03-01 18:18:37      Kingsbury St & Kinzie St     KA1503000043   
228495  2021-03-26 18:06:43                           NaN              NaN   

                   end_station_name end_station_id  start_lat  start_lng  \
0       Lincoln Ave & Diversey Pkwy          152.0  41.896400 -87.661000   
1                   Kosciuszko Park          499.0  41.924400 -87.715400   
2        Indiana Ave & Roosevelt Rd          255.0  41.894500 -87.617900   
3            Wood St & Augusta Blvd          657.0  41.903000 -87.697500   
4        Sheridan Rd & Lawrence Ave          323.0  41.890200 -87.626200   
...                             ...            ...        ...        ...   
228491         New St & Illinois St   TA1306000013  41.900960 -87.623777   
228492         New St & Illinois St   TA1306000013  41.889177 -87.638506   
228493        Clark St & Berwyn Ave   KA1504000146  41.900960 -87.623777   
228494         New St & Illinois St   TA1306000013  41.889177 -87.638506   
228495         New St & Illinois St   TA1306000013  41.880000 -87.630000   

          end_lat    end_lng member_casual  
0       41.932200 -87.658600        member  
1       41.930600 -87.723800        member  
2       41.867900 -87.623000        member  
3       41.899200 -87.672200        member  
4       41.969500 -87.654700        casual  
...           ...        ...           ...  
228491  41.890847 -87.618617        casual  
228492  41.890847 -87.618617        member  
228493  41.977997 -87.668047        member  
228494  41.890847 -87.618617        member  
228495  41.890818 -87.619540        member  

[3489539 rows x 13 columns]

In [12]:
#Checking the null values and counting with the sum() function
# Yeah i know it's weird the sum() function be counting
# I was thinking it should be the count() function

tripdataV1.isna().sum()


ride_id                    0
rideable_type              0
started_at                 0
ended_at                   0
start_station_name    122175
start_station_id      122801
end_station_name      143241
end_station_id        143702
start_lat                  0
start_lng                  0
end_lat                 4737
end_lng                 4737
member_casual              0
dtype: int64

Now we could see there are lot of null values such as:
1. start_station_name has 122175
2. start_station_id has 122801
3. end_station_name has 143241
4. end_station_id has 143702
5. end_lat has 4737
6. end_lng has 4737

We will handle this null values

In [13]:
# As we can see in my unique values i don't have nulls values

tripdataV1["ride_id"].isna().sum()


0

In [14]:
# Using this function to see this specific row index values

tripdataV1.iloc[228494]


ride_id                         E4CE369DF59D80C5
rideable_type                        docked_bike
started_at                   2020-05-12 13:14:43
ended_at                     2020-05-12 13:21:55
start_station_name    Larrabee St & Kingsbury St
start_station_id                              48
end_station_name           Dearborn St & Erie St
end_station_id                             110.0
start_lat                                41.8978
start_lng                               -87.6429
end_lat                                   41.894
end_lng                                 -87.6293
member_casual                             member
Name: 143718, dtype: object

## **Converting types**


In [15]:
# Checking again the number of rows

len(tripdataV1)


3489539

In [16]:
# Checking outliers

tripdataV1["member_casual"].nunique()


2

In [17]:
# Checking outliers

tripdataV1["rideable_type"].nunique()


3

In [18]:
# Watching a brief description about the column so i can be sure of the type

tripdataV1["started_at"].describe()



count                 3489539
unique                3040064
top       2020-07-12 13:16:59
freq                       12
Name: started_at, dtype: object

**As seem above the started_at is in an object data type**
**The ended_at column has the same format and we want them to be datetime**

In [19]:
# Converting to a datetime format the column that was in an object format


tripdataV1["started_at"] = pd.to_datetime(tripdataV1["started_at"])


/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  after removing the cwd from sys.path.


In [20]:
# Checking if the started_at column already has a datatime type format

tripdataV1["started_at"]


0        2020-04-26 17:45:14
1        2020-04-17 17:08:54
2        2020-04-01 17:54:13
3        2020-04-07 12:50:19
4        2020-04-18 10:22:59
                 ...        
228491   2021-03-20 14:58:56
228492   2021-03-02 11:35:10
228493   2021-03-09 11:07:36
228494   2021-03-01 18:11:57
228495   2021-03-26 17:58:14
Name: started_at, Length: 3489539, dtype: datetime64[ns]

In [21]:
# Converting the ended_at column to datetime format

tripdataV1["ended_at"] = pd.to_datetime(tripdataV1["ended_at"])



/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [22]:
# Checking if the changed was made

tripdataV1["ended_at"]


0        2020-04-26 18:12:03
1        2020-04-17 17:17:03
2        2020-04-01 18:08:36
3        2020-04-07 13:02:31
4        2020-04-18 11:15:54
                 ...        
228491   2021-03-20 17:22:47
228492   2021-03-02 11:43:37
228493   2021-03-09 11:49:11
228494   2021-03-01 18:18:37
228495   2021-03-26 18:06:43
Name: ended_at, Length: 3489539, dtype: datetime64[ns]

In [23]:
tripdataV1.info()

# Checking my data types format and we have two date times as intended

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3489539 entries, 0 to 228495
Data columns (total 13 columns):
 #   Column              Dtype         
---  ------              -----         
 0   ride_id             object        
 1   rideable_type       object        
 2   started_at          datetime64[ns]
 3   ended_at            datetime64[ns]
 4   start_station_name  object        
 5   start_station_id    object        
 6   end_station_name    object        
 7   end_station_id      object        
 8   start_lat           float64       
 9   start_lng           float64       
 10  end_lat             float64       
 11  end_lng             float64       
 12  member_casual       object        
dtypes: datetime64[ns](2), float64(4), object(7)
memory usage: 372.7+ MB


## Manipulating the data

* **Add new columns**

In [24]:
# Using the dt.dayofweek to find the day of the week for each date

tripdataV1["started_at"].dt.dayofweek

0         6
1         4
2         2
3         1
4         5
         ..
228491    5
228492    1
228493    1
228494    0
228495    4
Name: started_at, Length: 3489539, dtype: int64

In [25]:
# Using the dt.day_name() to get the name of the day to each date

tripdataV1["started_at"].dt.day_name()


0            Sunday
1            Friday
2         Wednesday
3           Tuesday
4          Saturday
            ...    
228491     Saturday
228492      Tuesday
228493      Tuesday
228494       Monday
228495       Friday
Name: started_at, Length: 3489539, dtype: object

In [26]:
# df['month_of_date'] = df['date_given'].dt.month

tripdataV1["started_at"].dt.month


0         4
1         4
2         4
3         4
4         4
         ..
228491    3
228492    3
228493    3
228494    3
228495    3
Name: started_at, Length: 3489539, dtype: int64

In [27]:
# Using the dat.month_name to see the months of each date
# later we will assign it into a new column

tripdataV1["started_at"].dt.month_name()


0         April
1         April
2         April
3         April
4         April
          ...  
228491    March
228492    March
228493    March
228494    March
228495    March
Name: started_at, Length: 3489539, dtype: object

In [28]:
# Now taking the year this "dt." is kind a a library to manipulate date

tripdataV1["started_at"].dt.year

0         2020
1         2020
2         2020
3         2020
4         2020
          ... 
228491    2021
228492    2021
228493    2021
228494    2021
228495    2021
Name: started_at, Length: 3489539, dtype: int64

In [29]:
# Extracting the specific day

tripdataV1["started_at"].dt.day

0         26
1         17
2          1
3          7
4         18
          ..
228491    20
228492     2
228493     9
228494     1
228495    26
Name: started_at, Length: 3489539, dtype: int64

**After wacthing the new values we need to create new columns we are going to create and insert the new columns**

In [30]:
# Adding the day as a new columns 

tripdataV1["weekday"] = tripdataV1["started_at"].dt.day_name() #the code for the new "weekday" column

tripdataV1["day"] = tripdataV1["started_at"].dt.day
tripdataV1["month"] = tripdataV1["started_at"].dt.month
tripdataV1["year"] = tripdataV1["started_at"].dt.year


tripdataV1

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value inst

ride_id  rideable_type          started_at  \
0       A847FADBBC638E45    docked_bike 2020-04-26 17:45:14   
1       5405B80E996FF60D    docked_bike 2020-04-17 17:08:54   
2       5DD24A79A4E006F4    docked_bike 2020-04-01 17:54:13   
3       2A59BBDF5CDBA725    docked_bike 2020-04-07 12:50:19   
4       27AD306C119C6158    docked_bike 2020-04-18 10:22:59   
...                  ...            ...                 ...   
228491  9397BDD14798A1BA    docked_bike 2021-03-20 14:58:56   
228492  BBBEB8D51AAD40DA   classic_bike 2021-03-02 11:35:10   
228493  637FF754DA0BD9E1   classic_bike 2021-03-09 11:07:36   
228494  F8F43A0B978A7A35   classic_bike 2021-03-01 18:11:57   
228495  3AE64EA5BF43CF72  electric_bike 2021-03-26 17:58:14   

                  ended_at            start_station_name start_station_id  \
0      2020-04-26 18:12:03                  Eckhart Park               86   
1      2020-04-17 17:17:03     Drake Ave & Fullerton Ave              503   
2      2020-04-01 18:08:36          McClurg Ct & Erie St              142   
3      2020-04-07 13:02:31  California Ave & Division St              216   
4      2020-04-18 11:15:54          Rush St & Hubbard St              125   
...                    ...                           ...              ...   
228491 2021-03-20 17:22:47         Michigan Ave & Oak St            13042   
228492 2021-03-02 11:43:37      Kingsbury St & Kinzie St     KA1503000043   
228493 2021-03-09 11:49:11         Michigan Ave & Oak St            13042   
228494 2021-03-01 18:18:37      Kingsbury St & Kinzie St     KA1503000043   
228495 2021-03-26 18:06:43                           NaN              NaN   

                   end_station_name end_station_id  start_lat  start_lng  \
0       Lincoln Ave & Diversey Pkwy          152.0  41.896400 -87.661000   
1                   Kosciuszko Park          499.0  41.924400 -87.715400   
2        Indiana Ave & Roosevelt Rd          255.0  41.894500 -87.617900   
3            Wood St & Augusta Blvd          657.0  41.903000 -87.697500   
4        Sheridan Rd & Lawrence Ave          323.0  41.890200 -87.626200   
...                             ...            ...        ...        ...   
228491         New St & Illinois St   TA1306000013  41.900960 -87.623777   
228492         New St & Illinois St   TA1306000013  41.889177 -87.638506   
228493        Clark St & Berwyn Ave   KA1504000146  41.900960 -87.623777   
228494         New St & Illinois St   TA1306000013  41.889177 -87.638506   
228495         New St & Illinois St   TA1306000013  41.880000 -87.630000   

          end_lat    end_lng member_casual    weekday  day  month  year  
0       41.932200 -87.658600        member     Sunday   26      4  2020  
1       41.930600 -87.723800        member     Friday   17      4  2020  
2       41.867900 -87.623000        member  Wednesday    1      4  2020  
3       41.899200 -87.672200        member    Tuesday    7      4  2020  
4       41.969500 -87.654700        casual   Saturday   18      4  2020  
...           ...        ...           ...        ...  ...    ...   ...  
228491  41.890847 -87.618617        casual   Saturday   20      3  2021  
228492  41.890847 -87.618617        member    Tuesday    2      3  2021  
228493  41.977997 -87.668047        member    Tuesday    9      3  2021  
228494  41.890847 -87.618617        member     Monday    1      3  2021  
228495  41.890818 -87.619540        member     Friday   26      3  2021  

[3489539 rows x 17 columns]

**We have added the new four columns**
**The weekday, day, month and year**
**This columns will help us make a deep analysis between the casual and members**

In [31]:
# calculating the ride lenght in seconds

tripdataV1["ride_length_s"] = (tripdataV1["ended_at"] - tripdataV1["started_at"]).dt.seconds/60 

tripdataV1["ride_length_s"]


/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


0          26.816667
1           8.150000
2          14.383333
3          12.200000
4          52.916667
             ...    
228491    143.850000
228492      8.450000
228493     41.583333
228494      6.666667
228495      8.483333
Name: ride_length_s, Length: 3489539, dtype: float64

In [32]:
# tripdataV1["weekday"] = tripdataV1["started_at"].dt.day_name() 

# tripdataV1["day"] = tripdataV1["started_at"].dt.day
# tripdataV1["month"] = tripdataV1["started_at"].dt.month
# tripdataV1["year"] = tripdataV1["started_at"].dt.year


In [33]:
# Now we are going to add new one column here
# the year_month
# .map(str)

tripdataV1["year_month"] = tripdataV1["year"].map(str) + " - " + tripdataV1["month"].map(str) + " (" + tripdataV1["started_at"].dt.month_name().map(str) + ")"

tripdataV1["year_month"]

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


0         2020 - 4 (April)
1         2020 - 4 (April)
2         2020 - 4 (April)
3         2020 - 4 (April)
4         2020 - 4 (April)
                ...       
228491    2021 - 3 (March)
228492    2021 - 3 (March)
228493    2021 - 3 (March)
228494    2021 - 3 (March)
228495    2021 - 3 (March)
Name: year_month, Length: 3489539, dtype: object

In [34]:
# Now we are going to add new one column here
# the dayofweek_weekday
# .map(str)
 
tripdataV1["dayofweek_weekday"] = tripdataV1["started_at"].dt.dayofweek.map(str) + " - " + tripdataV1["weekday"].map(str)

tripdataV1["dayofweek_weekday"]


/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


0            6 - Sunday
1            4 - Friday
2         2 - Wednesday
3           1 - Tuesday
4          5 - Saturday
              ...      
228491     5 - Saturday
228492      1 - Tuesday
228493      1 - Tuesday
228494       0 - Monday
228495       4 - Friday
Name: dayofweek_weekday, Length: 3489539, dtype: object

**Now we have added all the new columns we are going to need for our analysis**

In [35]:
# Watch my data again

tripdataV1



ride_id  rideable_type          started_at  \
0       A847FADBBC638E45    docked_bike 2020-04-26 17:45:14   
1       5405B80E996FF60D    docked_bike 2020-04-17 17:08:54   
2       5DD24A79A4E006F4    docked_bike 2020-04-01 17:54:13   
3       2A59BBDF5CDBA725    docked_bike 2020-04-07 12:50:19   
4       27AD306C119C6158    docked_bike 2020-04-18 10:22:59   
...                  ...            ...                 ...   
228491  9397BDD14798A1BA    docked_bike 2021-03-20 14:58:56   
228492  BBBEB8D51AAD40DA   classic_bike 2021-03-02 11:35:10   
228493  637FF754DA0BD9E1   classic_bike 2021-03-09 11:07:36   
228494  F8F43A0B978A7A35   classic_bike 2021-03-01 18:11:57   
228495  3AE64EA5BF43CF72  electric_bike 2021-03-26 17:58:14   

                  ended_at            start_station_name start_station_id  \
0      2020-04-26 18:12:03                  Eckhart Park               86   
1      2020-04-17 17:17:03     Drake Ave & Fullerton Ave              503   
2      2020-04-01 18:08:36          McClurg Ct & Erie St              142   
3      2020-04-07 13:02:31  California Ave & Division St              216   
4      2020-04-18 11:15:54          Rush St & Hubbard St              125   
...                    ...                           ...              ...   
228491 2021-03-20 17:22:47         Michigan Ave & Oak St            13042   
228492 2021-03-02 11:43:37      Kingsbury St & Kinzie St     KA1503000043   
228493 2021-03-09 11:49:11         Michigan Ave & Oak St            13042   
228494 2021-03-01 18:18:37      Kingsbury St & Kinzie St     KA1503000043   
228495 2021-03-26 18:06:43                           NaN              NaN   

                   end_station_name end_station_id  start_lat  start_lng  \
0       Lincoln Ave & Diversey Pkwy          152.0  41.896400 -87.661000   
1                   Kosciuszko Park          499.0  41.924400 -87.715400   
2        Indiana Ave & Roosevelt Rd          255.0  41.894500 -87.617900   
3            Wood St & Augusta Blvd          657.0  41.903000 -87.697500   
4        Sheridan Rd & Lawrence Ave          323.0  41.890200 -87.626200   
...                             ...            ...        ...        ...   
228491         New St & Illinois St   TA1306000013  41.900960 -87.623777   
228492         New St & Illinois St   TA1306000013  41.889177 -87.638506   
228493        Clark St & Berwyn Ave   KA1504000146  41.900960 -87.623777   
228494         New St & Illinois St   TA1306000013  41.889177 -87.638506   
228495         New St & Illinois St   TA1306000013  41.880000 -87.630000   

          end_lat    end_lng member_casual    weekday  day  month  year  \
0       41.932200 -87.658600        member     Sunday   26      4  2020   
1       41.930600 -87.723800        member     Friday   17      4  2020   
2       41.867900 -87.623000        member  Wednesday    1      4  2020   
3       41.899200 -87.672200        member    Tuesday    7      4  2020   
4       41.969500 -87.654700        casual   Saturday   18      4  2020   
...           ...        ...           ...        ...  ...    ...   ...   
228491  41.890847 -87.618617        casual   Saturday   20      3  2021   
228492  41.890847 -87.618617        member    Tuesday    2      3  2021   
228493  41.977997 -87.668047        member    Tuesday    9      3  2021   
228494  41.890847 -87.618617        member     Monday    1      3  2021   
228495  41.890818 -87.619540        member     Friday   26      3  2021   

        ride_length_s        year_month dayofweek_weekday  
0           26.816667  2020 - 4 (April)        6 - Sunday  
1            8.150000  2020 - 4 (April)        4 - Friday  
2           14.383333  2020 - 4 (April)     2 - Wednesday  
3           12.200000  2020 - 4 (April)       1 - Tuesday  
4           52.916667  2020 - 4 (April)      5 - Saturday  
...               ...               ...               ...  
228491     143.850000  2021 - 3 (March)      5 - Saturday  
228492       8.450000  2021 - 3 (March)  

## Converting and Saving to a CSV file format

**Now we are going to save the last version of our dataset and also export it into a CSV file.**

**It is cleaned now so, will be good to have it cleaned as a CSV**

In [36]:
tripdataV2 = tripdataV1


In [37]:
# df.to_csv("your_name.csv")

tripdataV2.to_csv("tripdata_clean.csv")


# Analyze Process

## Descriptive analysis

**Now we are going to have a deep qualitative and quantintative dept of our data**

In [38]:
# Checking the description of the ride_length_s

tripdataV2["ride_length_s"].describe()


count    3.489539e+06
mean     2.855945e+01
std      9.031763e+01
min      0.000000e+00
25%      7.950000e+00
50%      1.461667e+01
75%      2.683333e+01
max      1.439983e+03
Name: ride_length_s, dtype: float64

In [39]:

tripdataV2.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 3489539 entries, 0 to 228495
Data columns (total 20 columns):
 #   Column              Dtype         
---  ------              -----         
 0   ride_id             object        
 1   rideable_type       object        
 2   started_at          datetime64[ns]
 3   ended_at            datetime64[ns]
 4   start_station_name  object        
 5   start_station_id    object        
 6   end_station_name    object        
 7   end_station_id      object        
 8   start_lat           float64       
 9   start_lng           float64       
 10  end_lat             float64       
 11  end_lng             float64       
 12  member_casual       object        
 13  weekday             object        
 14  day                 int64         
 15  month               int64         
 16  year                int64         
 17  ride_length_s       float64       
 18  year_month          object        
 19  dayofweek_weekday   object        
dtypes: 

In [40]:
# a closer view to my numeric data

tripdataV2.describe()

start_lat     start_lng       end_lat       end_lng           day  \
count  3.489539e+06  3.489539e+06  3.484802e+06  3.484802e+06  3.489539e+06   
mean   4.190417e+01 -8.764494e+01  4.190444e+01 -8.764522e+01  1.586048e+01   
std    4.364222e-02  2.575990e-02  4.373714e-02  2.589148e-02  8.763003e+00   
min    4.164000e+01 -8.787000e+01  4.154000e+01 -8.807000e+01  1.000000e+00   
25%    4.188224e+01 -8.765888e+01  4.188266e+01 -8.765917e+01  8.000000e+00   
50%    4.190000e+01 -8.764170e+01  4.190068e+01 -8.764275e+01  1.600000e+01   
75%    4.193000e+01 -8.762773e+01  4.193120e+01 -8.762775e+01  2.300000e+01   
max    4.208000e+01 -8.752000e+01  4.216000e+01 -8.744000e+01  3.100000e+01   

              month          year  ride_length_s  
count  3.489539e+06  3.489539e+06   3.489539e+06  
mean   7.518390e+00  2.020107e+03   2.855945e+01  
std    2.569745e+00  3.096849e-01   9.031763e+01  
min    1.000000e+00  2.020000e+03   0.000000e+00  
25%    6.000000e+00  2.020000e+03   7.950000e+00  
50%    8.000000e+00  2.020000e+03   1.461667e+01  
75%    9.000000e+00  2.020000e+03   2.683333e+01  
max    1.200000e+01  2.021000e+03   1.439983e+03

## Analyze the key differences between members and casuals

**Now we are going to analyze the members and casuals ride_length_s**

In [41]:



tripdataV2.groupby("member_casual").sum()


start_lat     start_lng       end_lat       end_lng  \
member_casual                                                           
casual         5.993541e+07 -1.253602e+08  5.984433e+07 -1.251690e+08   
member         8.629083e+07 -1.804802e+08  8.618433e+07 -1.802572e+08   

                    day     month        year  ride_length_s  
member_casual                                                 
casual         22741159  10806300  2889421301   5.746084e+07  
member         32604609  15429414  4159822431   4.219847e+07

* **We have to create another dataframe to that we are going to call compare_ride_length**

**With this dataframe we are going to compare the differences between member and casuals by ride length in seconds**

In [42]:
# Comparison between member and casuals by ride_length_s


compare_ride_length = pd.DataFrame({'member_casual': tripdataV2["member_casual"],
                   'ride_length_s': tripdataV2["ride_length_s"]})

compare_ride_length

member_casual  ride_length_s
0             member      26.816667
1             member       8.150000
2             member      14.383333
3             member      12.200000
4             casual      52.916667
...              ...            ...
228491        casual     143.850000
228492        member       8.450000
228493        member      41.583333
228494        member       6.666667
228495        member       8.483333

[3489539 rows x 2 columns]

**After create the dataframe we can use the describe function to see the differences between members and casuals**

In [43]:
compare_ride_length.groupby("member_casual").describe()


ride_length_s                                                   \
                      count       mean        std  min        25%        50%   
member_casual                                                                  
casual            1430351.0  40.172548  94.778521  0.0  11.583333  21.266667   
member            2059188.0  20.492773  86.168355  0.0   6.533333  11.533333   

                                       
                     75%          max  
member_casual                          
casual         40.450000  1439.983333  
member         20.233333  1439.983333

* **We have to do another separeted comparison because the describe function don't show this to us.**
* **We will use the sum function to see the total length did between members and casuals**

In [44]:
# Using the sum function to see the total length between members and casulas

compare_ride_length.groupby("member_casual").sum()


ride_length_s
member_casual               
casual          5.746084e+07
member          4.219847e+07